src/main.cpp

```diff
@@ -150,13 +150,20 @@ void handle_options(quick_lint_js::options o) {
     quick_lint_js::run_lsp_server();
     std::exit(EXIT_SUCCESS);
   }
-  if (o.files_to_lint.empty()) {
+  if (o.files_to_lint.empty() && !o.stdinput) {
     std::cerr << "error: expected file name\n";
     std::exit(EXIT_FAILURE);
   }
 
   quick_lint_js::any_error_reporter reporter =
       quick_lint_js::any_error_reporter::make(o.output_format, &o.exit_fail_on);
+  if (o.stdinput) {
+    quick_lint_js::read_file_result source = quick_lint_js::read_stdin();
+    file_to_lint fstdin{.path = "<stdin>", .vim_bufnr = std::nullopt};
```

Comment:
Almost-blocking: I would expect `quick-lint-js --vim-file-bufnr=42 --stdin` to have `--vim-file-bufnr=42` apply to stdin. I think we need to process `--stdin` as if it was a file in `options::files_to_lint`.